docs/articles/clustering/cluster-metrics.md

```diff
@@ -0,0 +1,391 @@
+* `MemoryUsed` - total memory allocated to the currently running process
+* `MemoryAvailable` - memory, available for the process
+* `MaxMemoryRecommended` - if set, memory limit recommended for current process
+* `Processors` - number of available processors
+* `CpuProcessUsage` - CPU usage by current process
+* `CpuTotalUsage` - total CPU usage
+
+> Note: currently, due to some .NET Core limitations `CpuTotalUsage` is the same as `CpuProcessUsage` metrics, 
+> but this is something to be fixed in near future.
+
+Cluster metrics extension comes with built-in `Akka.Cluster.Metrics.Collectors.DefaultCollector` collector implementation, 
+which collects all metrics defined above.
+
+You can also plug-in your own metrics collector implementation.
+
+By default, metrics extension will use collector provider fall back and will try to load them in this order:
+1. configured user-provided collector (see `Configuration` section for details)
+2. built-in `Akka.Cluster.Metrics.Collectors.DefaultCollector` collector
+
+## Metrics Events
+
+Metrics extension periodically publishes current snapshot of the cluster metrics to the node system event bus.
+
+The publication interval is controlled by the `akka.cluster.metrics.collector.sample-interval` setting.
+
+The payload of the `Akka.Cluster.Metrics.Events.ClusterMetricsChanged` event will contain latest metrics of the node as well as 
+other cluster member nodes metrics gossip which was received during the collector sample interval.
+
+You can subscribe your metrics listener actors to these events in order to implement custom node lifecycle:
+```c#
+ClusterMetrics.Get(Sys).Subscribe(metricsListenerActor);
+```
+
+## Adaptive Load Balancing
+
+The `AdaptiveLoadBalancingPool` / `AdaptiveLoadBalancingGroup` performs load balancing of messages to cluster nodes based on the cluster metrics data. 
+It uses random selection of routees with probabilities derived from the remaining capacity of the corresponding node. 
+It can be configured to use a specific `IMetricsSelector` implementation to produce the probabilities, a.k.a. weights:
+
+* `memory` / `MemoryMetricsSelector` - Used and max available memory. Weights based on remaining memory capacity: (max - used) / max
+* `cpu` / `CpuMetricsSelector` - CPU utilization in percentage. Weights based on remaining cpu capacity: 1 - utilization
+* `mix` / `MixMetricsSelector` - Combines memory and cpu. Weights based on mean of remaining capacity of the combined selectors.
+
+The collected metrics values are smoothed with [exponential weighted moving average](https://en.wikipedia.org/wiki/Moving_average#Exponential_moving_average). 
+In the cluster configuration you can adjust how quickly past data is decayed compared to new data.
+
+Let’s take a look at this router in action. What can be more demanding than calculating factorials?
+
+The backend worker that performs the factorial calculation:
+
+```c#
+public class FactorialBackend : ReceiveActor
```

Comment:
We should include this type of material inside https://github.com/akkadotnet/akka.net/tree/dev/src/core/Akka.Docs.Tests and use DocFx syntax to link from the code in that project, i.e.:

https://raw.githubusercontent.com/akkadotnet/akka.net/dev/docs/articles/streams/stream-dynamic.md

```
[!code-csharp[KillSwitchDocTests.cs](../../../src/core/Akka.Docs.Tests/Streams/KillSwitchDocTests.cs?name=unique-shutdown)]
```

This guarantees that in the event that we refactor / change these APIs, the example included in the documentation will break - thus if we do things this way we can guarantee that all examples are statically type checked... And since they're modeled as unit tests, they also have to be functionally accurate as well. 

This is also designed to help prevent code samples in the documentation from rotting over time, since most contributors (myself included) don't go back over them and check them as often. So going forward we should implement this best practice throughout the Akka.NET documentation. Does that make sense?

Reply:
Sure, makes sense. I was thinking about something like that, but checked out cluster sharding sample and saw that code was just pasted there, so made in the same manner.

Thanks for samples and links, I will update

Reply:
Would you mind opening an issue to fix that with our Cluster.Sharding documentation too? We should, ideally, go back and replace these static code samples with the dynamically linked ones everywhere.

Reply:
Sure, but it is not only Cluster.Sharding - there are few more files to be fixed. 

Here is an issue: https://github.com/akkadotnet/akka.net/issues/4169